### PR TITLE
Book category setup 

### PR DESCRIPTION
**Pull request by vsych321**

### CODE REVIEW

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Pls resolve the merge conflicts

File: src/main/java/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+@RequiredArgsConstructor
+@RestController
+@Validated
+@RequestMapping("/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/controller/CategoryController.java]:
```suggestion
    @ResponseStatus(HttpStatus.CREATED)
```

File: src/main/java/bookstore/entity/Book.java [hunk 2]
```diff
@@ -38,6 +46,18 @@ public class Book {
     private String description;
     private String coverImage;
 
-    @Column(nullable = false)
+    @Column(nullable = false, name = "is_deleted")
     private boolean isDeleted = false;
+
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Book.java]: LAZY default type for ManyToMany relation

File: src/main/java/bookstore/entity/Book.java [hunk 3]
```diff
@@ -38,6 +46,18 @@ public class Book {
...
+    @JoinTable(
+            name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id")
+    )
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    private Set<Category> categories = new HashSet<>();
+
+
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Book.java]:
```suggestion
```

File: src/main/java/bookstore/entity/Category.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+@Where(clause = "is_deleted=false")
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
+    private String name;
+
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Category.java:27]: In request it required parameter

File: src/main/java/bookstore/repository/bookrepo/BookRepository.java [hunk 5]
```diff
@@ -1,10 +1,16 @@
...
+import java.util.List;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+
+    //@EntityGraph(attributePaths = "categories")
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/bookrepo/BookRepository.java]: Let’s  remove the commented code

File: src/main/java/bookstore/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -1,34 +1,44 @@
...
+    @Transactional
+    public BookDtoWithoutCategoryIds save(CreateBookRequestDto requestDto) {
         Book book = bookMapper.toBook(requestDto);
-        return bookMapper.toDto(bookRepository.save(book));
+        setCategories(book,requestDto.categoryIds());
+        return bookMapper.toDtoWithoutCategories(bookRepository.save(book));
     }
 
     @Override
+    @Transactional
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/impl/BookServiceImpl.java]: Let’s  use readOnly parameter

File: src/main/java/bookstore/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -37,6 +47,7 @@ public List<BookDto> findAll(Pageable pageable) {
...
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/impl/BookServiceImpl.java]: the same

File: src/main/java/bookstore/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -59,11 +72,27 @@ public List<BookDto> searchBookByParams(BookSearchParametersDto parameters, Page
...
         Book book = bookRepository.findById(id).orElseThrow(() -> new EntityNotFoundException(
                 "Can't update book by id " + id)
         );
         bookMapper.updateBook(requestDto,book);
+        setCategories(book, requestDto.categoryIds());
         return bookMapper.toDto(bookRepository.save(book));
     }
+
+    @Override
+    @Transactional
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/impl/BookServiceImpl.java]: Do you really need Transactional for each method?

File: src/main/resources/db/changelog/changes/05-create-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                  name: description
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/05-create-categories-table.yaml]: add one empty line  (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes)

File: src/main/java/bookstore/entity/Book.java [hunk 10]
```diff
@@ -38,6 +45,18 @@ public class Book {
...
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Book.java]: remove redundant  lines

File: src/main/java/bookstore/repository/bookrepo/BookRepository.java (hunk 5 above)
Comment (liliia-ponomarenko) [reply src/main/java/bookstore/repository/bookrepo/BookRepository.java]: not fixed

File: src/main/java/bookstore/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -29,7 +29,7 @@ public UserResponseDto register(UserRegistrationRequestDto request) {
         }
         User user = userMapper.toUser(request);
         user.setRoles(Collections.singleton(roleRepository
-                .findRoleByName(Role.RoleName.ROLE_USER)));
+                .findRoleByName(Role.RoleName.ROLE_ADMIN)));
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/impl/UserServiceImpl.java:32]: Why did you change role here? Should each user have an admin role?

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 12]
```diff
@@ -6,4 +6,10 @@ databaseChangeLog:
...
   - include:
-      file: db/changelog/changes/04-create-user_role-table.yaml
\ No newline at end of file
+      file: db/changelog/changes/04-create-user_role-table.yaml
+  - include:
+      file: db/changelog/changes/05-create-categories-table.yaml
+  - include:
+      file: db/changelog/changes/06-create-book_category-table.yaml
+  - include:
+      file: db/changelog/changes/07-insert-roles-to-db.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: add empty line

Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/07-insert-roles.yaml
```

File: src/test/java/bookstore/OnlineBookStoreApplicationTests.java [hunk 13]
```diff
@@ -1,13 +1,13 @@
...
 
 @SpringBootTest
 class OnlineBookStoreApplicationTests {
 
-    @Test
-    void contextLoads() {
-    }
+//    @Test
+//    void contextLoads() {
+//    }
```
Comment (liliia-ponomarenko) [inline src/test/java/bookstore/OnlineBookStoreApplicationTests.java]: remove commented code

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Be attentive! Let’s fix all mistakes ;)

File: src/main/java/bookstore/service/impl/UserServiceImpl.java (hunk 11 above)
Comment (vsych321) [reply src/main/java/bookstore/service/impl/UserServiceImpl.java:32]: Anyway, I'm changing roles in DB

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊
